src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.hpp

```diff
@@ -4,13 +4,28 @@
 
 #pragma once
 
-#include <openvino/core/node.hpp>
 #include <ngraph/runtime/host_tensor.hpp>
-#include "static_shape.hpp"
-
+#include <openvino/core/core.hpp>
+#include <openvino/core/node.hpp>
 
+#include "static_shape.hpp"
 
 void shape_inference(ov::Node* op,
                      const std::vector<ov::StaticShape>& input_shapes,
                      std::vector<ov::StaticShape>& output_shapes,
-                     const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {});
\ No newline at end of file
+                     const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {});
+
+class IShapeInfer {
+public:
+    virtual void infer(const std::vector<ov::StaticShape>& input_shapes,
+                       std::vector<ov::StaticShape>& output_shapes,
+                       const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data) = 0;
+
+    // infer may generate padding as by-product, these APIs is designed to retrieve them back
+    virtual const ov::CoordinateDiff& get_pads_begin();
+    virtual const ov::CoordinateDiff& get_pads_end();
+
+    virtual ngraph::Node * get_op();
```

Comment:
The use of `opToShapeInfer` reference was a reluctant decision, since we had to call `validate_and_ifer_types()` and retrieve auto pads somehow. But, we do not want to have an access to the ngraph operation on the nodes implementation level until it is strictly necessary. It looks like this ShapeInfer interface allows us to hide the ngraph node reference from its users. It seems the only place where `get_op()` call is used is the input shapes collection routine. So, m.b. it is better to add `infer` method without `input_shapes` so they will be retrieved automatically? And, it looks like the `infer` method can return `output_shapes` instead of filling up the user provided vector.

Reply:
OK, get_op() is removed and `output_shapes` changed as `infer()` return type.

As for `input_shapes`, I don't fully understand your proposal, but I believe shape inference API should be plugin independent, so it seems to be more reasonable for caller collect input shapes and feed them into shape inference API.

Reply:
The idea is to have two `infer()` methods: the first one with input shapes, provided by the user, the second one without input shapes so they are collected from the ngraph op. For me it looks more logical than have an interface that returns these input shapes just for passing them to the infer method, looks like an extra step.

Reply:
Yes, I was wrong, the ngraph op was used to get a number of the inputs/outputs. But I thought that input shapes were also retrived from the op. Sorry.